### PR TITLE
Java monitoring metrics 

### LINKED ISSUE #209: Add stats export for ledger writer

### PR DESCRIPTION
**Pull request by WilliamG240**
This fixes #209, #210, #211 and adds Java monitoring metrics exporting to Cloud Monitoring.

Important metrics:
- custom/jvm/.
  - java virtual memory metrics such as garbage collection, memory usage, threads 
- custom/hikaricp/connection
   - Database connection pool usage time (how long a database connection takes)
- custom/cache/.
   - Guava cache metrics such as hits,misses,puts
- custom/system/.
   - System metrics
- custom/process/uptime
   - Uptime of the Java virtual machine
- custom/hibernate/.
   - Hibernate datasource metrics such as entity load times, slowest query time, number of collection updates, etc.

### CODE REVIEW

File: src/ledgerwriter/src/test/java/anthos/samples/bankofanthos/ledgerwriter/LedgerWriterControllerTest.java [hunk 1]
```diff
@@ -73,7 +80,26 @@
     @BeforeEach
     void setUp() {
         initMocks(this);
+        StackdriverMeterRegistry meterRegistry = new StackdriverMeterRegistry(new StackdriverConfig() {
```
Comment (murog) [inline src/ledgerwriter/src/test/java/anthos/samples/bankofanthos/ledgerwriter/LedgerWriterControllerTest.java:83]: will this need to be added for balancereader as well?

File: src/ledgerwriter/src/main/resources/application.properties [hunk 2]
```diff
@@ -5,3 +5,4 @@ java.util.logging.manager=org.apache.logging.log4j.jul.LogManager
 spring.sleuth.sampler.probability=1.0
 spring.sleuth.web.skipPattern=(^cleanup.*|.+favicon.*)
 spring.sleuth.log.slf4j.enabled=false
+server.tomcat.mbeanregistry.enabled=true
```
Comment (murog) [inline src/ledgerwriter/src/main/resources/application.properties:8]: Why was this needed?

File: src/ledgerwriter/src/test/java/anthos/samples/bankofanthos/ledgerwriter/LedgerWriterControllerTest.java (hunk 1 above)
Comment (WilliamG240) [reply src/ledgerwriter/src/test/java/anthos/samples/bankofanthos/ledgerwriter/LedgerWriterControllerTest.java:83]: Yes! Just did!

File: src/ledgerwriter/src/main/resources/application.properties (hunk 2 above)
Comment (WilliamG240) [reply src/ledgerwriter/src/main/resources/application.properties:8]: This allows all tomcat metrics to be recorded

File: src/balancereader/src/test/java/anthos/samples/bankofanthos/balancereader/BalanceReaderControllerTest.java [hunk 3]
```diff
@@ -63,7 +74,28 @@
...
+
+            @Override
+            @Nullable
+            public String get(String key) {
+                return null;
+            }
+        }, clock);
+
+        when(cache.stats()).thenReturn(new CacheStats(0,
+            0,0,0,0,0));
```
Comment (askmeegs) [inline src/balancereader/src/test/java/anthos/samples/bankofanthos/balancereader/BalanceReaderControllerTest.java:96]: just curious, what do all these 0s represent? are they required for initialization? is there a default value for the CacheStats object?

File: src/ledgerwriter/src/main/java/anthos/samples/bankofanthos/ledgerwriter/LedgerWriterApplication.java [hunk 4]
```diff
@@ -72,4 +77,43 @@ public RestTemplate restTemplate() {
     public void destroy() {
         LOGGER.info("LedgerWriter service shutting down");
     }
+
+    /**
+     * Initializes Meter Registry with custom Stackdriver configuration
+     *
+     * @return the StackdriverMeterRegistry with configuration
+     */
+    @Bean
```
Comment (askmeegs) [inline src/ledgerwriter/src/main/java/anthos/samples/bankofanthos/ledgerwriter/LedgerWriterApplication.java:86]: is it possible to share this Bean across the 3 java services? or do you think it's more trouble than it's worth?

File: src/balancereader/src/main/resources/log4j2.xml [hunk 5]
```diff
@@ -23,6 +23,9 @@
         </Console>
     </Appenders>
     <Loggers>
+        <Logger name="org.hibernate" level="WARN" additivity="false">
```
Comment (askmeegs) [inline src/balancereader/src/main/resources/log4j2.xml:26]: based on this line add, I'm guessing we weren't logging hibernate plugin logs before this? is this to log if any metrics error happens?

File: src/balancereader/src/test/java/anthos/samples/bankofanthos/balancereader/BalanceReaderControllerTest.java (hunk 3 above)
Comment (WilliamG240) [reply src/balancereader/src/test/java/anthos/samples/bankofanthos/balancereader/BalanceReaderControllerTest.java:96]: There isn't a default value for the Cache Stats object and this is required since in the initialization of the balanceReader, the guavacachemetrics calls cache.stats().hitcount(), cache.stats().misscount(),etc and since the cache is a Mock the cache.stats() originally returned null which caused a nullPointerException.

File: src/balancereader/src/main/resources/log4j2.xml (hunk 5 above)
Comment (WilliamG240) [reply src/balancereader/src/main/resources/log4j2.xml:26]: Yes, we were previously not logging the hibernate plugin logs, and this would log any errors in the hibernate plugin (if an entity didnt load properly or if metrics error)

File: src/balancereader/src/test/java/anthos/samples/bankofanthos/balancereader/BalanceReaderControllerTest.java (hunk 3 above)
Comment (WilliamG240) [reply src/balancereader/src/test/java/anthos/samples/bankofanthos/balancereader/BalanceReaderControllerTest.java]: Turned the cache stats object into a mock to clean things up!

File: src/balancereader/src/main/java/anthos/samples/bankofanthos/balancereader/BalanceReaderController.java [hunk 6]
```diff
@@ -64,14 +63,16 @@
     @Autowired
     public BalanceReaderController(LedgerReader reader,
         JWTVerifier verifier,
+        StackdriverMeterRegistry meterRegistry,
         LoadingCache<String, Long> cache,
         @Value("${LOCAL_ROUTING_NUM}") final String localRoutingNum,
-        @Value("${VERSION") final String version) {
+        @Value("${VERSION}") final String version) {
```
Comment (murog) [inline src/balancereader/src/main/java/anthos/samples/bankofanthos/balancereader/BalanceReaderController.java:69]: nice catch 👍

File: src/balancereader/src/main/java/anthos/samples/bankofanthos/balancereader/BalanceReaderApplication.java [hunk 7]
```diff
@@ -66,4 +71,43 @@ public static void main(String[] args) {
     public void destroy() {
         LOGGER.info("BalanceReader service shutting down");
     }
+
+    /**
+     * Initializes Meter Registry with custom Stackdriver configuration
```
Comment (murog) [inline src/balancereader/src/main/java/anthos/samples/bankofanthos/balancereader/BalanceReaderApplication.java:76]: thanks for adding
